### DISCUSSION TITLE
Re-building and improving GPT-2

### DISCUSSION THREAD

**Discussion by garg-aayush** [karpathy/build-nanogpt#106] [Show and tell]
Over the last couple of weeks, I followed @karpathy’s [‘Let’s Reproduce GPT-2’](https://www.youtube.com/watch?v=l8pRSuU81PU) video religiously—making notes, implementing the logic line by line, and completing a re-implementation of GPT-2 from scratch.

I went a few steps further by implementing some of the improvements suggested by @karpathy (such as learning rate adjustments and data loader fixes), along with modern enhancements like RoPE and SwiGLU-FFN. My best-performing experiment `gpt2-rope`, achieved a validation loss of 2.987 and a HellaSwag accuracy of 0.320.

<img width="4800" height="1800" alt="gpt-2-implement-compare" src="https://github.com/user-attachments/assets/5db45bc2-1ef0-4e0a-b182-9c5a00cdf90f" />

| Experiment | Min Validation Loss | Max HellaSwag Acc | Description |
|------------|---------------------|-------------------|-------------|
| gpt2-baseline | 3.065753 | 0.303724 | Original GPT-2 architecture |
| gpt2-periodicity-fix | 3.063873 | 0.305517 | Fixed data loading periodicity |
| gpt2-lr-inc | 3.021046 | 0.315475 | Increased learning rate by 3x and reduced warmup steps |
| gpt2-global-datafix | 3.004503 | 0.316869 | Used global shuffling with better indexing |
| gpt2-rope | **2.987392** | **0.320155** | Replaced learned embeddings with RoPE |
| gpt2-swiglu<sup>*</sup>  | 3.031061 | 0.317467 | Replaced FFN with SwiGLU-FFN activation |

I really loved the whole process of writing the code, running multiple trainings and gradually seeing the losses improve. I learnt so much about LLMs pre-training from this single video. Honestly, the $200 I spent on compute over these two weeks was the best money I’ve spent lately. Learned a ton and had fun.

I have made sure to log everything, the code, training runs, checkpoints, notes:
- Repo: https://github.com/garg-aayush/building-from-scratch/blob/main/gpt-2/
- Runs: https://wandb.ai/garg-aayush/pre-training
- Dataset (training and validation): [Google Drive](https://drive.google.com/drive/folders/1FGHKpY0_jJmSR_j7ki4oyoxK-fJgldgG?usp=sharing)
- Best checkpoints for each experiment: [Google Drive](https://drive.google.com/drive/folders/1S9mFDMG3ZPjA-JGdx_814T_NCVSjUJO-?usp=sharing)
